caddyhttp/basicauth/basicauth.go

```diff
@@ -63,7 +63,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
 			// user; this replaces the request with a wrapped instance
-			r = r.WithContext(context.WithValue(r.Context(),
+			*r = *r.WithContext(context.WithValue(r.Context(),
```

Comment:
@jamesboswell That's all fine and well, but I strongly want to avoid creating special cases, which this certainly is, for there will only be more and more. Can we generalize the fix instead?